### PR TITLE
[LLVMGPU] Improve reduction lowering to do group reduction

### PR DESCRIPTION
**Pull request by ThomasRaoux**
Change to do vector distribution on a group instead of doing it on a
warp. That allows us to use extra parallelism. This also simplifies the
flow since we don't need to do a second level of tiling and distribute.

### CODE REVIEW

**Comment by ThomasRaoux** [comment]
depends on https://reviews.llvm.org/D129387

File: compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp [hunk 1]
```diff
@@ -51,6 +51,34 @@ static Value warpReduction(Location loc, OpBuilder &builder, Value input,
...
 }
 
+/// Emit reduction across a group for a given input.
+static Value groupReduction(Location loc, OpBuilder &builder, Value input,
+                            vector::CombiningKind kind, uint32_t size) {
+  const int warpSize = 32;
+  Value laneVal = warpReduction(loc, builder, input, kind, warpSize);
+  // if we have more than one warp, reduce across warps.
+  if (size > warpSize) {
+    uint32_t numWarp = size / warpSize;
```
Comment (okkwon) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp:61]: Is `size` always a multiple of 32?

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp:61]: Right now yes due to restriction for the compilation pipeline. But let me add an assert.

**Comment by okkwon** [comment]
Some tests were unsuccessful? Is your PR ready?

**Comment by ThomasRaoux** [comment]
Review greatly appreciated :)

File: compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp [hunk 2]
```diff
@@ -111,22 +139,23 @@ struct LLVMGPUReduceToGPUPass
       (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
     }
 
+    auto workgroupSize = llvm::to_vector<4>(llvm::map_range(
+        getEntryPoint(funcOp)->workgroup_size().getValue(),
+        [&](Attribute attr) { return attr.cast<IntegerAttr>().getInt(); }));
+    assert(workgroupSize[1] == 1 && workgroupSize[2] == 1);
```
Comment (okkwon) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp:148]: What made this guaranteed? Can you add a comment?

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp:148]: the KernelConfig.cpp code.

Comment (okkwon) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp:148]: So, I believe the meaning of workgroup size is different from the workgroup count at the flow level? At the flow level, the workgroup count shows the total number.

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp:148]: workgroup size is the number of threads per workgroup. That's what we care about here. This is not the same as workgroup count or workload per work indeed.
Here we want the number of threads to know how to distribute the work.

Comment (okkwon) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp:148]: Thanks for the explanation. The words are so similar to each other but representing a different concept.

**Review by okkwon — APPROVED** [review]

**Pull request merged**